src/metabase/models/field_usage.clj

```diff
@@ -0,0 +1,100 @@
+   :filter_args            mi/transform-json})
+
+(defn- filter->field-usage
+  [query stage filter-clause]
+  (let [filter-parts (lib/filter-parts query stage filter-clause)
+        field-id     (-> filter-parts :column :id)]
+    (when (int? field-id)
+      {:field_id    field-id
+       :used_in     :filter
+       :filter_op   (-> filter-parts :operator :short)
+       :filter_args (:args filter-parts)})))
+
+(defn- aggregation->field-usage
+  [query stage aggregation-clause]
+  (let [aggregation-column (lib/aggregation-column query stage aggregation-clause)
+        field-id           (:id aggregation-column)]
+    (when (int? field-id)
+      {:field_id             field-id
+       :used_in              :aggregation
+       :aggregation_function (first aggregation-clause)})))
+
+(defn- breakout->field-usage
+  [query stage breakout-clause]
+  (let [breakout-column (lib/breakout-column query stage breakout-clause)
+        field-id        (:id breakout-column)]
+    (when (int? field-id)
+      {:field_id               field-id
+       :used_in                :breakout
+       :breakout_temporal_unit (lib/raw-temporal-bucket breakout-clause)
+       :breakout_binning       (-> (lib/binning breakout-clause)
+                                   (select-keys [:strategy :bin-width :num-bins])
+                                   not-empty)})))
+
+(defn- expression->field-usage
+  [expresison-clause]
+  (when-let [field-ids (seq (lib.util.match/match expresison-clause [:field _ (id :guard int?)] id))]
+    (for [field-id field-ids]
+      {:field_id field-id
+       :used_in  :expression})))
+
+(defmulti ^:private pquery->field-usages-method
+  "Impl for [[pquery->field-usages]]."
+  {:arglists '([query stage-number x])}
+  (fn [_query _stage-number x] (lib.dispatch/dispatch-value x))
+  :hierarchy lib.hierarchy/hierarchy)
+
+(defmethod pquery->field-usages-method :mbql/query
+  [query _stage-number x]
+  (flatten (for [stage (range (lib/stage-count x))]
+                (pquery->field-usages-method query stage (lib.util/query-stage query stage)))))
+
+(defmethod pquery->field-usages-method :metabase.lib.stage/stage
+  [query stage-number x]
+  (concat
+    (keep #(filter->field-usage query stage-number %) (lib/filters query stage-number))
+    (keep #(aggregation->field-usage query stage-number %) (lib/aggregations query stage-number))
+    (keep #(breakout->field-usage query stage-number %) (lib/breakouts query stage-number))
+    (flatten (keep expression->field-usage (lib/expressions query stage-number)))
+    (flatten (keep #(pquery->field-usages-method query stage-number %) (:joins x)))))
+
+(defmethod pquery->field-usages-method :mbql/join
+  [query _stage-number x]
+  (let [join-query (fetch-source-query/resolve-source-cards (assoc query :stages (get x :stages)))]
+    ;; treat the source query as a :mbql/query
+    (pquery->field-usages-method join-query nil join-query)))
+
+(mu/defn pquery->field-usages
```

Comment:
naming suggestions are welcomed! I called this `pquery` to distinguish between normal query and mlv2 query aka pipelined mbql.

Reply:
I feel like just `query` vs `legacy-query` would be good names for mlv2 and mlv1 respectively. Otherwise just `mlv2-query` would be so much clearer that it's worth the verbosity.

Reply:
from the [MLv2 primer](https://www.notion.so/metabase/MBQL-Lib-v2-Primer-b89b9fba2d894f1db759554452bba354?pvs=4#be18da1d03e1420b9395bf610cee4704):

> “MLv2 query” is a slightly-incorrect shorthand for a pMBQL query, to distinguish it from the “legacy query” form.

We often use `sql` as a name for a SQL query, so maybe we should use `pmbql` as a name for a pMBQL query

Reply:
pmbql it is 3eaa54e01b